neqo-transport/src/tracking.rs

```diff
@@ -550,34 +540,22 @@ impl ::std::fmt::Display for RecvdPackets {
     }
 }
 
-#[derive(Debug)]
 pub struct AckTracker {
-    /// This stores information about received packets in *reverse* order
-    /// by spaces.  Why reverse?  Because we ultimately only want to keep
-    /// `ApplicationData` and this allows us to drop other spaces easily.
-    spaces: SmallVec<[RecvdPackets; 1]>,
+    spaces: EnumMap<PacketNumberSpace, Option<RecvdPackets>>,
 }
 
 impl AckTracker {
     pub fn drop_space(&mut self, space: PacketNumberSpace) {
-        let sp = match space {
-            PacketNumberSpace::Initial => self.spaces.pop(),
-            PacketNumberSpace::Handshake => {
-                let sp = self.spaces.pop();
-                self.spaces.shrink_to_fit();
-                sp
-            }
-            PacketNumberSpace::ApplicationData => panic!("discarding application space"),
-        };
-        assert_eq!(sp.unwrap().space, space, "dropping spaces out of order");
+        self.spaces[space].take();
+        assert_ne!(
```

Comment:
It's probably best to put this assertion before the destructive change.

There is another assertion:

```rust
if space == PacketNumberSpace::Handshake {
  assert!(self.spaces[PacketNumberSpace::Initial].is_none());
}
```